okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java

```diff
@@ -139,32 +140,33 @@ void onStreamSentBytes(int numBytes) {
     onSentBytes(numBytes);
   }
 
+  /**
+   * Must be called with holding the transport lock.
+   */
   public void transportHeadersReceived(List<Header> headers, boolean endOfStream) {
-    synchronized (lock) {
-      if (endOfStream) {
-        transportTrailersReceived(Utils.convertTrailers(headers));
-      } else {
-        transportHeadersReceived(Utils.convertHeaders(headers));
-      }
+    if (endOfStream) {
+      transportTrailersReceived(Utils.convertTrailers(headers));
+    } else {
+      transportHeadersReceived(Utils.convertHeaders(headers));
     }
   }
 
-  /**
-   * We synchronized on "lock" for delivering frames and updating window size, because
-   * the {@link #request(int)} call can be called in other thread for delivering frames.
-   */
   public void transportDataReceived(okio.Buffer frame, boolean endOfStream) {
```

Comment:
Hmm... I think I see. `finishStream()` can't be called with the lock held, because `stopIfNecessary()` will block waiting for the `frameWriter` to close() so it can close the socket. Maybe we should do the `socket.close()` on the writing thread?

Note that it seems we can't easily _not_ call `finishStream()` with without the lock held because `transportTrailersReceived()` can call `finishStream()`.